File: scala-experiment/src/main/scala/Parsers.scala
case class MatchResult[T](
    result: T,
    remainder: String
)

type Matcher[T] = String => Option[MatchResult[T]]

def stringLiteral(s: String, caseSensitive: Boolean = true): Matcher[String] =
    input =>
        val hasPrefix =
            if caseSensitive then input.startsWith(s)
            else input.toLowerCase().startsWith(s.toLowerCase())
        if hasPrefix then
            Some(
              MatchResult(
                result = input.substring(0, s.length()),
                remainder = input.substring(s.length())
              )
            )
        else None

extension (s: String)
    def toMatcher: Matcher[String] =
        stringLiteral(s)

def charRange(lower: Char, upper: Char): Matcher[Char] =
    input =>
        input.headOption.match
            case Some(value) if value >= lower && value <= upper =>
                Some(
                  MatchResult(result = value, remainder = input.substring(1))
                )
            case _ => None

def list[T](m: Matcher[T]*): Matcher[List[T]] =
    m.toList.match
        case head :: next =>
            val matchTail = list(next*)
            input =>
                head(input).match
                    case Some(MatchResult(headResult, remainder)) =>
                        matchTail(remainder).match
                            case Some(MatchResult(tailResult, remainder)) =>
                                Some(MatchResult(result = headResult :: tailResult, remainder))
                            case None =>
                                None
                    case None =>
                        None
        case Nil =>
            input => Some(MatchResult(result = List(), remainder = input))

extension [T](l: List[Matcher[T]])
    def toListMatcher: Matcher[List[T]] =
        list(l*)

def tuple2[T1, T2](m1: Matcher[T1], m2: Matcher[T2]): Matcher[(T1, T2)] =
    input =>
        m1(input).match
            case Some(MatchResult(r1, remainder)) =>
                m2(remainder).match
                    case Some(MatchResult(r2, remainder)) =>
                        Some(MatchResult((r1, r2), remainder))
                    case None => None
            case None => None

extension [T1, T2](m: (Matcher[T1], Matcher[T2]))
    def toMatcher: Matcher[(T1, T2)] =
        val (m1, m2) = m
        tuple2(m1, m2)

def tuple3[T1, T2, T3](m1: Matcher[T1], m2: Matcher[T2], m3: Matcher[T3]): Matcher[(T1, T2, T3)] =
    input =>
        m1(input).match
            case Some(MatchResult(r1, remainder)) =>
                m2(remainder).match
                    case Some(MatchResult(r2, remainder)) =>
                        m3(remainder).match
                            case Some(MatchResult(r3, remainder)) =>
                                Some(MatchResult((r1, r2, r3), remainder))
                            case None => None
                    case None => None
            case None => None

extension [T1, T2, T3](m: (Matcher[T1], Matcher[T2], Matcher[T3]))
    def toMatcher: Matcher[(T1, T2, T3)] =
        val (m1, m2, m3) = m
        tuple3(m1, m2, m3)

/*
TODO anyOf
def anyOf[T](m: Matcher[T]*): Matcher[List[T]] =

TODO optional
TODO atLeastZero
TODO atLeastOne
 */

File: ocaml-experiment/lib/parsers.ml
open Strings

type 'a match_result = { result : 'a; remainder : string } [@@deriving show]
type 'a matcher = string -> 'a match_result option

let match_literal (s : string) ?(case_sensitive : bool = true) : string matcher
    =
 fun input ->
  let has_prefix =
    if case_sensitive then String.starts_with input ~prefix:s
    else
      String.starts_with
        (String.lowercase_ascii input)
        ~prefix:(String.lowercase_ascii s)
  in
  if has_prefix then
    let result = String.sub input 0 (String.length s) in
    let remainder =
      String.sub input (String.length s) (String.length input - String.length s)
    in
    Some { result; remainder }
  else None

let match_char_range (lower : char) (upper : char) : char matcher =
 fun input ->
  match string_uncons input with
  | Some result, remainder ->
      if result >= lower && result <= upper then Some { result; remainder }
      else None
  | None, _ -> None

let rec match_seq (matchers : 'a matcher list) : 'a list matcher =
  match matchers with
  (* base case, no matchers, input passed through unchanged *)
  | [] -> fun input -> Some { result = []; remainder = input }
  | m :: matchers -> (
      (* a matcher that handles the 2nd and on *)
      let match_rest = match_seq matchers in
      fun input ->
        (* apply the first matcher *)
        match m input with
        | Some { result = result_head; remainder } -> (
            (* apply the rest of the matchers *)
            match match_rest remainder with
            | Some { result = result_tail; remainder } ->
                Some { result = result_head :: result_tail; remainder }
            | None -> None)
        | None -> None)

let match_seq2 (m1 : 't1 matcher) (m2 : 't2 matcher) : ('t1 * 't2) matcher =
 fun input ->
  match m1 input with
  | Some { result = r1; remainder } -> (
      match m2 remainder with
      | Some { result = r2; remainder } -> Some { result = (r1, r2); remainder }
      | None -> None)
  | None -> None

let match_seq3 (m1 : 't1 matcher) (m2 : 't2 matcher) (m3 : 't3 matcher) :
    ('t1 * 't2 * 't3) matcher =
 fun input ->
  match m1 input with
  | Some { result = r1; remainder } -> (
      match m2 remainder with
      | Some { result = r2; remainder } -> (
          match m3 remainder with
          | Some { result = r3; remainder } ->
              Some { result = (r1, r2, r3); remainder }
          | None -> None)
      | None -> None)
  | None -> None

(*
TODO match any of
TODO match optional
TODO match in range
TODO match at least zero
TODO match at least one
TODO ignore prefix
TODO ignore suffix
TODO ignore prefix and suffix
TODO match integers
TODO match floats
TODO match any whitespace
*)
